lib/puppet/provider/package/pip.rb

```diff
@@ -116,99 +118,148 @@ def query
     self.class.validate_command(command)
 
     Dir.mktmpdir("puppet_pip") do |dir|
       command_and_options = [command, 'install', "#{@resource[:name]}", '-d', "#{dir}", '-v']
       command_and_options << install_options if @resource[:install_options]
+
       execpipe command_and_options do |process|
         process.collect do |line|
           # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
+          if line =~ /Using version .+? \(newest of versions: (.+?)\)/
+            version_list = $1.split(', ').sort do |a,b|
+              PipVersion.compare(a,b)
+            end
+            return version_list
           end
         end
         return nil
       end
     end
   end
 
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PipVersion.new(version)
+      included_available_versions.push(version) if should_range.include?(version)
+    end
+
+    included_available_versions.last
+  end
+
   # Install a package.  The ensure parameter may specify installed,
   # latest, a version number, or, in conjunction with the source
   # parameter, an SCM revision.  In that case, the source parameter
   # gives the fully-qualified URL to the repository.
-
   def install
     command = resource_or_provider_command
     self.class.validate_command(command)
 
+    should = @resource[:ensure]
     command_options = %w{install -q}
     command_options +=  install_options if @resource[:install_options]
     if @resource[:source]
-      if String === @resource[:ensure]
-        command_options << "#{@resource[:source]}@#{@resource[:ensure]}#egg=#{@resource[:name]}"
+      if String === should
+        command_options << "#{@resource[:source]}@#{should}#egg=#{@resource[:name]}"
       else
         command_options << "#{@resource[:source]}#egg=#{@resource[:name]}"
       end
     else
-      case @resource[:ensure]
-      when String
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
-      when :latest
+      if should == :latest
         command_options << "--upgrade" << @resource[:name]
       else
-        command_options << @resource[:name]
+        begin
+          should_range = VersionRange.parse(should, PipVersion)
+          should = best_version(should_range)
+
+          if should
+            command_options << "#{@resource[:name]}==#{should}"
+            @resource[:ensure] = should.to_s
+          end
+        rescue VersionRange::ValidationFailure, PipVersion::ValidationFailure
+          Puppet.debug("Cannot parse #{should} as a pip version range, falling through")
+          
+          if should.is_a? String
+            command_options << "#{@resource[:name]}==#{should}"
+          else
+            command_options << @resource[:name]
+          end
+        end
```

Comment:
I believe an approach similar to the regex match here https://github.com/puppetlabs/puppet/pull/8030/files#diff-0df6fa18d947b3857abe3ff1c68dfd44R318 would be better

I think it's more readable to keep the switch and do something like:

```ruby
when :latest ... # (do latest stuff)
when /[<>=]/ ... # (do version range stuff)
when String ... # (do normal version stuff)
else ... # (do no version stuff)
```

Reply:
I don't think provider should know about how a range is defined.
It should try to parse `should` as a Range and rescue ValidationError

Reply:
But then you would try to parse a normal version like `1.2.3` when you don't actually want to.

Reply:
so, after chatting we concluded that we shouldn't do a `/[<>=]/` check, only an `is_a?(String)` check before attempting to parse as a version range.

Reply:
the same thing applies to the parsing inside the `insync` method